app/forms/candidate_interface/reference/email_address_step.rb

```diff
@@ -0,0 +1,69 @@
+module CandidateInterface
+  module Reference
+    class EmailAddressStep < DfE::Wizard::Step
+      include Rails.application.routes.url_helpers
+
+      attr_accessor :email_address
+      delegate :reference_process, :reference, :return_to_path,
+               :application_choice, to: :wizard
+
+      validates :email_address, presence: true,
+                                valid_for_notify: true,
+                                length: { maximum: 100 }
+
+      validates :reference, presence: true
+
+      validate :email_address_unique
+      validate :email_address_not_own
+
+      def self.permitted_params
+        [:email_address]
+      end
+
+      def previous_step
+        return_to_path || candidate_interface_references_name_path(
+          reference_process,
+          reference.referee_type.dasherize,
+          reference&.id,
+          application_id: application_choice&.id,
+        )
+      end
+
+      def next_step
+        return_to_path || process_next_path
+      end
+
+    private
+
+      def process_next_path
+        if reference_process == 'accept-offer'
```

Comment:
We are checking 9 times the same equalities. It might be difficult to someone that is not me and you to understand all the possibilities of the reference process.

I think is worth discussion an simple abstraction

## Suggestion one: Module

```ruby
module ReferenceProcess
  extend ActiveSupport::Concern

  included do
    REFERENCE_PROCESSES = {
      candidate_details: 'candidate-details',
      accept_offer: 'accept-offer',
      request_references: 'request-references'
    }.freeze

    REFERENCE_PROCESSES.each do |key, value|
      define_method("#{key}?") do
        reference_process == value
      end
    end
  end

  class_methods do
    def reference_processes
      REFERENCE_PROCESSES
    end
  end
end
```
Usage:
```ruby
class SomeClass # or component etc
  include ActiveModel::Model
  include ReferenceProcess

  attr_accessor :reference_process

  def some_method
    if candidate_details?
      # do something for candidate-details
    elsif accept_offer?
      # do something for accept-offer
    elsif request_references?
      # do something for request-references
    end
  end
end
```

## Suggestion 2: A simple class

```ruby
class ReferenceProcess
  CANDIDATE_DETAILS = 'candidate-details'
  ACCEPT_OFFER = 'accept-offer'
  REQUEST_REFERENCES = 'request-references'

  PROCESSES = {
    candidate_details: CANDIDATE_DETAILS,
    accept_offer: ACCEPT_OFFER,
    request_references: REQUEST_REFERENCES
  }.freeze

  attr_reader :reference_process

  def initialize(reference_process)
    @reference_process = reference_process
  end

  PROCESSES.each do |key, value|
    define_method("#{key}?") do
      @reference_process == value
    end
  end

  def self.processes
    PROCESSES
  end
end
```
Usage:
```
class Reference
  attr_reader :reference_process

  def initialize(reference_process)
    @reference_process = ReferenceProcess.new(reference_process)
  end

  def some_method
    if @reference_process.candidate_details?
      # do something for candidate-details
    elsif @reference_process.accept_offer?
      # do something for accept-offer
    elsif @reference_process.request_references?
      # do something for request-references
    end
  end
end
```

I don't have any particular, as long we could 1. Understand quickly all the process, centralize somewhere the checks and just call in the 9 necessary places.

Reply:
Like the module approach, I'll have a go at implementing it